File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
import sys
from collections import deque
from collections import defaultdict

n, m = map(int, sys.stdin.readline().split())


def bfs(start, target):
    que = deque()
    que.append([start, 0])
    dx = [-1, 1, 2]
    answer = abs(target - start)
    visit = defaultdict(bool)
    visit[start] = True
    while que:
        cur_pos, cur_cnt = que.popleft()
        if cur_pos == target:
            answer = answer if cur_cnt > answer else cur_cnt
        for i in range(3):
            next_pos = cur_pos + dx[i] if i < 2 else cur_pos * dx[i]
            if 0 <= next_pos <= target * 2 and not visit[next_pos]:
                que.append([next_pos, cur_cnt + 1])
                visit[next_pos] = True

    return answer

print(bfs(n, m))

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class Main {

	private static int n, m;
	public static void main(String[] args) {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		try {
			Integer[] arr = Arrays.stream(br.readLine().split(" ")).map(Integer::parseInt).toArray(Integer[]::new);
			n = arr[0]; m = arr[1];
			int answer = Math.abs(n - m);
			
			System.out.print(Math.min(bfs(), answer));
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	private static int bfs() {
		Queue<Integer[]> que = new LinkedList<>();
		boolean[] visit = new boolean[200001];
		int[] dx = {2, 1, -1};
		que.add(new Integer[] {n, 0});
		while(!que.isEmpty()) {
			Integer[] arr = que.poll();
			int cur = arr[0], cnt = arr[1];
			for(int i=0 ; i<3 ; i++) {
				int value = cur + dx[i];
				if(i == 0) value = cur * dx[i];
				if(value >= 0 && value < 200000 && !visit[value]) {
					if(value == m)return cnt + 1;
					que.add(new Integer[] {value, cnt + 1});
					visit[value] = true;
					}
				}
			}
		return 0;
	}
}
